cmake/IncludeGrpc.cmake

```diff
@@ -0,0 +1,120 @@
+# Copyright 2018 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# gRPC always requires thread support.
+find_package(Threads)
+include(${CMAKE_CURRENT_LIST_DIR}/PkgConfigHelper.cmake)
+
+# Configure the gRPC dependency, this can be found as a submodule, package, or
+# installed with pkg-config support.
+set(GOOGLE_CLOUD_CPP_GRPC_PROVIDER "module"
+        CACHE STRING "How to find the gRPC library")
+set_property(CACHE GOOGLE_CLOUD_CPP_GRPC_PROVIDER
+        PROPERTY STRINGS "module" "package" "vcpkg" "pkg-config")
+
+if ("${GOOGLE_CLOUD_CPP_GRPC_PROVIDER}" STREQUAL "module")
+    if (NOT GRPC_ROOT_DIR)
+        set(GRPC_ROOT_DIR ${CMAKE_CURRENT_SOURCE_DIR}/third_party/grpc)
+    endif ()
+    if (NOT EXISTS "${GRPC_ROOT_DIR}/CMakeLists.txt")
+        message(ERROR "GOOGLE_CLOUD_CPP_GRPC_PROVIDER is \"module\" but GRPC_ROOT_DIR is wrong")
+    endif ()
+    add_subdirectory(${GRPC_ROOT_DIR} third_party/grpc EXCLUDE_FROM_ALL)
+    add_library(gRPC::grpc++ ALIAS grpc++)
+    add_library(gRPC::grpc ALIAS grpc)
+    add_library(protobuf::libprotobuf ALIAS libprotobuf)
+
+    # The binary name is different on some platforms, use CMake magic to get it.
+    set(PROTOBUF_PROTOC_EXECUTABLE $<TARGET_FILE:protoc>)
+    mark_as_advanced(PROTOBUF_PROTOC_EXECUTABLE)
+    set(PROTOC_GRPCPP_PLUGIN_EXECUTABLE $<TARGET_FILE:grpc_cpp_plugin>)
+    mark_as_advanced(PROTOC_GRPCPP_PLUGIN_EXECUTABLE)
+elseif ("${GOOGLE_CLOUD_CPP_GRPC_PROVIDER}" STREQUAL "package"
+        OR "${GOOGLE_CLOUD_CPP_GRPC_PROVIDER}" STREQUAL "vcpkg")
+    find_package(GRPC REQUIRED grpc>=1.8)
```

Comment:
Here Required grpc version is >=1.8 and in first CMakeList.txt file required grpc version is >=1.4
Same for protobuf version here >=3.5 and in first CMakeLists.txt file is > 3.0
Is it correct?

Reply:
Yes.  gRPC does not compile with protobuf < 3.4.   I should actually update the gRPC version to 1.9, we ran into crashing bugs before that version.